pkg/engine/executor/pipeline.go

```diff
@@ -120,3 +121,102 @@ func emptyPipeline() Pipeline {
 		return Exhausted
 	})
 }
+
+// prefetchWrapper wraps a [Pipeline] with pre-fetching capability,
+// reading data in a separate goroutine to enable concurrent processing.
+type prefetchWrapper struct {
+	Pipeline // the pipeline that is wrapped
+
+	initialized bool // internal state to indicate whether the pre-fetching goroutine is running
+
+	ch    chan state // the results channel for pre-fetched items
+	state state      // internal state representing the last pre-fetched item
+
+	ctx    context.Context         // context to control lifecycle
+	cancel context.CancelCauseFunc // cancellation function for the context
+}
+
+var _ Pipeline = (*prefetchWrapper)(nil)
+
+// newPrefetchingPipeline creates a new prefetching pipeline wrapper that reads data from the underlying pipeline
+// in a separate goroutine. This allows for concurrent processing where data can be fetched ahead of time
+// while the consumer is processing the current batch.
+//
+// The prefetching pipeline maintains a buffered channel with capacity 1 to store the next batch,
+// enabling pipeline parallelism and potentially improving throughput.
+//
+// The function accepts a [context.Context] `ctx` and a [Pipeline] `p`, which is the underlying pipeline to wrap
+// with pre-fetching capability.
+//
+// Returns a [prefetchWrapper] that implements the [Pipeline] interface.
+func newPrefetchingPipeline(ctx context.Context, p Pipeline) *prefetchWrapper {
+	ctx, cancel := context.WithCancelCause(ctx)
+	return &prefetchWrapper{
+		Pipeline: p,
+		ch:       make(chan state),
+		ctx:      ctx,
+		cancel:   cancel,
+	}
+}
+
+// Read implements [Pipeline].
+func (p *prefetchWrapper) Read() error {
+	p.init()
+	return p.read()
+}
+
+func (p *prefetchWrapper) init() {
+	if p.initialized {
+		return
+	}
+
+	p.initialized = true
+	go p.prefetch(p.ctx) // nolint:errcheck
+}
+
+func (p prefetchWrapper) prefetch(ctx context.Context) error {
+	// Close channel on exit
+	defer close(p.ch)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			var s state
+			s.err = p.Pipeline.Read()
+			if s.err != nil {
+				p.ch <- s
+				return s.err
+			}
+			s.batch, _ = p.Pipeline.Value()
```

Comment:
Should this also set `s.err`? (And then should we return if it does provide a non-nil error to cancel the goroutine?) 

Without setting `s.err`, there's an edge case where a child pipeline doesn't return an error for the `Read` call, but _does_ return an error on the `Value` call.